src/v/config/throughput_control_group.cc

```diff
@@ -106,57 +148,118 @@ ss::sstring throughput_control_group::validate() const {
     if (unlikely(name != noname && contains_control_character(name))) {
         return "Group name contains invalid character";
     }
-    if (unlikely(client_id_re && !client_id_re->ok())) {
-        return ss::format("Invalid client_id regex. {}", client_id_re->error());
+    if (!client_id_matcher) {
+        return {};
     }
-    return {};
+    return std::visit(
+      [](auto&& v) -> ss::sstring {
+          using T = std::decay_t<decltype(v)>;
+          if constexpr (std::is_same_v<
+                          T,
+                          client_id_matcher_type::unspecified_tag>) {
+              return {};
+          } else if constexpr (std::is_same_v<T, copyable_RE2>) {
+              if (likely(v.ok())) {
+                  return {};
+              } else {
+                  return ss::format("Invalid client_id regex. {}", v.error());
+              }
+          } else {
+              static_assert(
+                always_false_v<T>, "non-exhaustive client_id_re_type visitor");
+          }
+      },
+      client_id_matcher->v);
 }
 
 } // namespace config
 
 namespace YAML {
 
 Node convert<config::throughput_control_group>::encode(const type& tcg) {
+    using namespace config;
     Node node;
-    if (tcg.name != config::noname) {
-        node[config::ids::name] = tcg.name;
+
+    if (tcg.name != noname) {
+        node[ids::name] = tcg.name;
     }
-    if (tcg.client_id_re) {
-        node[config::ids::client_id] = tcg.client_id_re->pattern();
+
+    if (tcg.client_id_matcher) {
+        YAML::Node client_id_node = node[ids::client_id];
+        std::visit(
+          [&client_id_node](auto&& v) {
+              using T = std::decay_t<decltype(v)>;
+              if constexpr (std::is_same_v<
+                              T,
+                              client_id_matcher_type::unspecified_tag>) {
+                  client_id_node = fmt::format(
+                    "{}{}", selector_prefix, selector_empty);
+              } else if constexpr (std::is_same_v<T, copyable_RE2>) {
+                  client_id_node = v.pattern();
+              } else {
+                  static_assert(
+                    always_false_v<T>,
+                    "non-exhaustive client_id_re_type visitor");
+              }
```

Comment:
if you stick with the std::variant, doesn't std::visit allow you to provide overrides for each type in the variant so that you don't have to do the type inspection with constexpr and have this static_assert(false)?